### ISSUE TITLE
Code Splitting PoC

### ISSUE THREAD

**Issue by gabemart** [akvo/akvo-lumen#300]
We should used [code splitting](https://webpack.github.io/docs/code-splitting.html) to avoid delivering the entire app on first-load. This will improve first-load performance.

This issue will track a proof-of-concept implementation of code splitting for lumen.

**Comment by gabemart** [comment]
I suggest the following approach:

# Application

I suggest we split the three main containers - `Dataset`, `Visualisation` and `Dashboard` - into separate bundles. They are well isolated from each other, so there isn't a lot of code duplication between bundles. It seems likely that at least some users will only need to access one or two of these containers in a given session, so loading on-demand seems like a performance win.

I also suggest that, because the dependencies for visualisations themselves are quite large and are needed in both `Visualisation` and `Dashboard`, we split these into a separate `ChartPreload` bundle. This bundle is pre-loaded before loading either the `Visualisation` or `Dashboard` containers. This means that in either the visualisation or dashboard editor, all visualisations are guaranteed to be immediately available, but we avoid loading the visualisation dependences when loading the Dataset editor, where no visualisations are needed.

# Public

For public visualisations and dashboards, I suggest we split by visualisation dependency, so that `rd3` is bundled with `DashBarChart`, `DashPieChart` etc. and `reactLeaflet` is bundled with `DashMap`.

(I still need to remove the outdated "Dash" prefix from these components, but that will be easiest to do once the initial code-splitting PR is merged)

Given that shared visualisations and dashboards will be viewed on a much wider range of devices and connection speeds than the main application, performance is critical here, so I think it makes sense to have separate bundles for separate visualisation libraries, whereas in the main application we bundle all visualisation dependencies together.

# Bundle sizes
````
current, dev
============
# app
app.bundle.js  4.34 MB

# public
pub.bundle.js  1.81 MB

current, production (minified)
=============
# app
app.bundle.js  1.69 MB

# public
pub.bundle.js   685 kB

Code-splitting, dev
==============
# app
app.bundle.js                   2.85 MB
1.DashChartPreload.bundle.js    909 kB
2.Visualisation.bundle.js       49.4 kB
3.Dataset.bundle.js             304 kB
4.Dashboard.bundle.js           244 kB
//                              -------
// TOTAL                        4.16 MB

# public
pub.bundle.js                   793 kB
6.reactLeaflet.bundle.js        671 kB       
7.rd3.bundle.js                 640 kB
//                              -------
// Total                        2.01MB
// Note - Total for "public" is larger than in develop, but only because public dashboards are
// broken in develop (Immutable.js is now needed in the viewer, but is not included)

Code-splitting, production (minified);
===============
# app
app.bundle.js                 1.04 MB
0.DashChartPreload.bundle.js   434 kB
7.Visualisation.bundle.js     23.1 kB
4.Dataset.bundle.js            113 kB
6.Dashboard.bundle.js         84.7 kB
//                            -------
// TOTAL                      1.62 MB

# public
pub.bundle.js                  232 kB
2.reactLeaflet.bundle.js       286 kB
5.rd3.bundle.js                282 kB
//                             ------
// Total                       800 kB
````

# Load Times

**Comment by mtwestra** [comment]
@gabe, what is the status of this issue?

**Comment by gabemart** [comment]
The initial work is now in develop, and any future code splitting can be dealt with in a new issue.
